### ISSUE TITLE
[RFC][API-Unstable] Intel GPU distributed Backend integration in `torch-xpu-ops`and registeration in PyTorch

### ISSUE THREAD

**Issue by zhangxiaoli73** [pytorch/pytorch#141741]
### 🚀 The feature, motivation and pitch

# Motivation

This Request for Comments (RFC) document aims to propose and discuss Intel GPU distributed support in PyTorch. This initiative begins with Intel distributed backend (`XCCL`) integration into PyTorch component `torch-xpu-ops`, and registration in PyTorch distributed Python package. 
The RFC outlines a high-level design strategy for this integration. NOTE: the device name for Intel GPU in PyTorch is XPU. Therefore, `XCCL` represents XPU collective communications library in this RFC.

# Design
**1.	Intel GPU distributed Backend integration in PyTorch `torch-xpu-ops`**

In the current design, PyTorch distributed utilizes c10d::ProcessGroup class as an interface to manage multiple communication backends (inherited from c10::Backend) and provide collective APIs that can be dispatched based on device type and backend. 

Regarding per-backend implementation, c10d::ProcessGroupNCCL targets the CUDA device with backend name “nccl”. Similarly, we would like add c10d::ProcessGroupXCCL on Intel GPU device with new backend name “xccl”. We can visualize this design as below.
 
![P1](https://github.com/user-attachments/assets/b38e9e22-50b2-41d5-83cb-71154ea9f55d)

Regarding code structure, XCCL backend source code will be put in PyTorch [torch-xpu-ops](https://github.com/intel/torch-xpu-ops), while those code will be built into `libtorch_xpu.so`.

![P2](https://github.com/user-attachments/assets/2edf7168-edfb-4d3c-bf47-2ace4d8f76f5)

**2.	Intel distributed Backend register in PyTorch distributed package**

All XPU components in PyTorch are on par with CUDA in PyTorch, illustrated in chart below. For example, `libtorch.so` supports both CUDA and XPU stream/device, `libtorch_cuda.so` contains CUDA ATen and collective ops while libtorch_xpu.so contains ATen and collective ops for XPU. 
![P3](https://github.com/user-attachments/assets/3893f49a-1ffb-4338-9d7c-d77bed213a1c)

Consequently, we expect XCCL backend could handle Python binding and backend register in the same way as NCCL backend.
1)	Add `ProcessGroupXCCL` Python module binding in [torch/csrc/distributed/c10d/init.cpp](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/init.cpp), and those code will be built in `libtorch_python.so`.
2)	Register `XCCL` name and XCCL backend in PyTorch Python Backend, including distributed [backend_type_map](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L268), [backend_capability](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L261) and [default_device_backend_map](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L256).
3)	Add `XCCL` name in native ProcessGroup [backend type list](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroup.hpp#L74).

 
# PR Plan
The code changes involve some parts of PyTorch. To be clear and concise, we will split those changes into 3 PRs for easy to review with below priority.
- [x] Backend ProcessGroupXCCL and collective `allreduce` as an entrypoint to `torch-xpu-ops`.
- [x] Backend ProcessGroupXCCL Python binding in PyTorch and register `XCCL` in PyTorch distributed Python package.
- [x] Remaining collectives to `torch-xpu-ops`.

cc @gujinghui @EikanWang @fengyuan14 @guangyey

**Comment by zhangxiaoli73** [comment]
@gujinghui @kwen2501 @wconstab Could you please help to review this RFC?

**Comment by wconstab** [comment]
haven't reviewed code yet, but just wanted to comment that the updated design and PR plan looks good to me and matches my understanding from the previous  discussions.

**Comment by kwen2501** [comment]
LGTM.
Regarding
> Add ProcessGroupXCCL Python module binding in [torch/csrc/distributed/c10d/init.cpp](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/init.cpp), and those code will be built in libtorch_python.so.

Maybe we can think where would be the best place -- if needed -- to put the binding.

**Comment by zhangxiaoli73** [comment]
> LGTM. Regarding
> 
> > Add ProcessGroupXCCL Python module binding in [torch/csrc/distributed/c10d/init.cpp](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/init.cpp), and those code will be built in libtorch_python.so.
> 
> Maybe we can think where would be the best place -- if needed -- to put the binding.

Thanks for your comments. Let's follow what we discuss in PR #141856 - keep it as is and then start to think how to refine the registration process. .

**Comment by zhangxiaoli73** [comment]
> Couple questions:
> 
>     * To which degree `gloo` distributed backend is/will be supported for XPU? It seems to be a recommended fallback option for CUDA's `nccl` in some places in documentation.
> 
>     * `c10d::allgather_,` `c10d::_allgather_base_`, etc. ops are currently not implemented for XPU backend. Will their implementation work across both newly proposed `xccl` and existing `gloo`? (I see from experiment that these are implemented for CUDA). On this topic, please, see [xpu: torch distributed ops like c10d::allgather are not implemented for XPU backend #142096](https://github.com/pytorch/pytorch/issues/142096) I submitted earlier today.

Hi, regarding your questions:
1) `gloo` is not supported on XPU. Currently, only `xccl` distributed backend could work on XPU device.
2) I think if you use `xccl` (you could expect `xccl` backend available after 3 tasks done recorded in this RFC), then those collectives could work correctly on XPU device.

**Comment by zhangxiaoli73** [comment]
Update this RFC according to requirements from RFC new process https://github.com/pytorch/pytorch/issues/152134:

- [ ] Unit Tests: basic distributed tests in https://github.com/intel/torch-xpu-ops/tree/main/test/xpu/distributed
- [ ] CI Coverage: https://github.com/pytorch/pytorch/pull/150927 and https://github.com/intel/torch-xpu-ops/pull/1409 and https://github.com/intel/torch-xpu-ops/pull/1288
- [ ] Doc String
- [ ] Tutorial:

**Comment by anitakat** [comment]
Hi @zhangxiaoli73 ! Thanks so much for commenting with the info relevant to the release. Could I ask that you please create a **Release Information** section in the description and add those there under "release 2.8"? It'll make it easier for the release team to find. 
Thanks so much.

**Comment by zhangxiaoli73** [comment]
Add Release Information section for this RFC:

**Release information** 

In PyTorch 2.8

Proposal of Marketing/Blog text post below:

[API-Unstable] Intel GPU distributed backend (XCCL) in PyTorch

XCCL is a distributed backend for Intel GPUs which allows users to enable various distributed training paradigms like DDP (DistributedDataParallel), FSDP (FullyShardedDataParallel), PP (pipeline parallelism), and TP (tensor parallelism) on XPU devices. XCCL backend provides all communication ops defined in PyTorch, such as `allreduce`, `allgather`, and `reducescatter`. XCCL backend can be applied transparently on XPU devices or explicitly specified with “xccl" name during PyTorch process group initialization.

CC: @EikanWang @gujinghui 
CC: @anitakat

**Comment by anitakat** [comment]
Thanks @zhangxiaoli73. Any tutorials for this yet?

**Comment by zhangxiaoli73** [comment]
@anitakat Yes, we submitted some docs PR for this XCCL backend. It would be great if you can help review and comment.
https://github.com/pytorch/pytorch/pull/155834
https://github.com/pytorch/pytorch/pull/155497
